File: Example/LuaSample/SampleEpubFilePlugin.lua
-- C++ 会在 init/run 前把当前 EpubTranslator 注入为全局 luaTranslator。
-- EPUB 的解包、JSON 生成和回打包仍由 C++ 侧负责，Lua 这里演示如何插入生命周期并查看映射表。

local function countMap(map)
    local count = 0
    for _ in pairs(map) do
        count = count + 1
    end
    return count
end

local function logEpubJsonInfo()
    local jsonToInfoMap = luaTranslator.m_jsonToInfoMap
    utils.logger:info("EPUB JSON 映射数量: " .. tostring(countMap(jsonToInfoMap)))

    for jsonPath, jsonInfo in pairs(jsonToInfoMap) do
        utils.logger:info(string.format(
            "EPUB JSON: %s, HTML: %s, textNodes: %d",
            tostring(jsonPath),
            tostring(jsonInfo.htmlPath),
            #jsonInfo.metadata
        ))
    end
end

local function processCurrentFilesWithLua()
    if luaTranslator.m_transEngine == TransEngine.DumpName then
        luaTranslator.m_controller:updateBar(luaTranslator.m_controller.m_totalSentences)
        return
    end

    if luaTranslator.m_transEngine == TransEngine.NameTrans then
        luaTranslator.m_nameTranslator:run(luaTranslator.m_nameTablePath)
        return
    end

    if luaTranslator.m_transEngine == TransEngine.GenDict then
        luaTranslator.m_dictionaryGenerator:generate(luaTranslator.m_projectDir / "ProjGptDict-Gen.toml")
        return
    end

    local relFilePaths = luaTranslator.m_currentRunRelFilePaths
    if relFilePaths == nil then
        return
    end

    luaTranslator:normalJsonProcessFiles(relFilePaths)

    if luaTranslator.m_reuseRepeatedBlocks and luaTranslator.m_transEngine ~= TransEngine.ShowNormal then
        luaTranslator:resolveRepeatedBlockReferences()
    end
    if luaTranslator.m_agentEnabled and luaTranslator.m_transAgent ~= nil then
        luaTranslator.m_transAgent:applyAgentSuggestions()
    end
end

function init()
    utils.logger:info("SampleEpubFilePlugin 初始化")
    utils.logger:info("当前项目目录: " .. tostring(luaTranslator.m_projectDir))
end

function run()
    luaTranslator:normalJsonInit()
    luaTranslator:epubInit()
    luaTranslator:epubBeforeRun()
    luaTranslator:normalJsonBeforeRun()

    local ok, err = pcall(function()
        logEpubJsonInfo()
        processCurrentFilesWithLua()
    end)

    luaTranslator:normalJsonAfterRun()
    if not ok then
        error(err)
    end
end

function runStandardLifecycle()
    luaTranslator:normalJsonInit()
    luaTranslator:epubInit()
    luaTranslator:epubBeforeRun()
    luaTranslator:normalJsonBeforeRun()
    luaTranslator:normalJsonProcess()
    luaTranslator:normalJsonAfterRun()
end

function unload()
    utils.logger:info("SampleEpubFilePlugin 卸载")
end

File: Example/PythonSample/SampleEpubFilePlugin.py
import gpp_plugin_api as gpp

import threading
from pathlib import Path
from typing import Callable, cast


pythonTranslator = cast(gpp.EpubTranslator, None)
logger = cast(gpp.spdlogLogger, None)


def init() -> None:
    logger.info("SampleEpubFilePlugin 初始化")
    logger.info("当前项目目录: " + str(pythonTranslator.m_projectDir))


def run() -> None:
    pythonTranslator.normalJsonInit()
    pythonTranslator.epubInit()
    pythonTranslator.epubBeforeRun()

    originOnFileProcessed = cast(Callable[[object], None], pythonTranslator.m_onFileProcessed)

    def onFileProcessed(relFilePath: object) -> None:
        logger.info("EPUB 片段处理完成: " + str(relFilePath))
        originOnFileProcessed(relFilePath)

    # m_onFileProcessed 是 EPUB 判断所有 json 是否完成并回打包的关键回调。
    # 覆盖成 Python 闭包后，需用 Python 线程逐文件调用 processFile，避免 C++ 线程池直接调用 Python 闭包。
    # 另，调用可能同时发生，如有需要请在 onFileProcessed 内上锁 (originOnFileProcessed 内部是线程安全的)
    pythonTranslator.m_onFileProcessed = onFileProcessed

    pythonTranslator.normalJsonBeforeRun()
    try:
        processCurrentFilesWithPythonThreads()
    finally:
        pythonTranslator.normalJsonAfterRun()


def processCurrentFilesWithPythonThreads() -> None:
    controller = pythonTranslator.m_controller

    if pythonTranslator.m_transEngine == gpp.TransEngine.DumpName:
        controller.updateBar(controller.m_totalSentences)
        return

    if pythonTranslator.m_transEngine == gpp.TransEngine.NameTrans:
        nameTranslator = pythonTranslator.m_nameTranslator
        if nameTranslator is None:
            raise RuntimeError("NameTranslator 未创建")
        nameTranslator.run(pythonTranslator.m_nameTablePath)
        return

    if pythonTranslator.m_transEngine == gpp.TransEngine.GenDict:
        dictionaryGenerator = pythonTranslator.m_dictionaryGenerator
        if dictionaryGenerator is None:
            raise RuntimeError("DictionaryGenerator 未创建")
        dictionaryGenerator.generate(pythonTranslator.m_projectDir / "ProjGptDict-Gen.toml")
        return

    relFilePaths = pythonTranslator.m_currentRunRelFilePaths
    if relFilePaths is None:
        return

    maxWorkers = min(pythonTranslator.m_threadsNum, len(relFilePaths))

    nextFileIndex = 0
    nextFileIndexLock = threading.Lock()
    stopEvent = threading.Event()
    errors: list[BaseException] = []
    errorsLock = threading.Lock()

    def worker(threadId: int) -> None:
        nonlocal nextFileIndex
        while not stopEvent.is_set() and not controller.shouldStop():
            with nextFileIndexLock:
                if nextFileIndex >= len(relFilePaths):
                    return
                relFilePath = Path(relFilePaths[nextFileIndex])
                nextFileIndex += 1

            try:
                processOneFile(relFilePath, threadId)
            except BaseException as e:
                with errorsLock:
                    errors.append(e)
                stopEvent.set()
                return

    threads = [
        threading.Thread(target=worker, args=(threadId,), name=f"gppPyEpub{threadId}")
        for threadId in range(maxWorkers)
    ]
    for thread in threads:
        thread.start()
    for thread in threads:
        thread.join()

    if errors:
        raise errors[0]

    if (pythonTranslator.m_reuseRepeatedBlocks
            and pythonTranslator.m_transEngine != gpp.TransEngine.ShowNormal):
        pythonTranslator.resolveRepeatedBlockReferences()
    if pythonTranslator.m_agentEnabled and pythonTranslator.m_transAgent is not None:
        pythonTranslator.m_transAgent.applyAgentSuggestions()


def processOneFile(relFilePath: Path, threadId: int) -> None:
    controller = pythonTranslator.m_controller
    if controller.shouldStop():
        return

    controller.addThreadNum()
    try:
        pythonTranslator.processFile(relFilePath, threadId)
    finally:
        controller.reduceThreadNum()


def unload() -> None:
    logger.info("SampleEpubFilePlugin 卸载")
